flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java

```diff
@@ -640,6 +640,74 @@ public SubtaskState answer(InvocationOnMock invocation) throws Throwable {
 		verify(rawOperatorStateHandle).discardState();
 	}
 
+	/**
+	 * FLINK-5985
+	 *
+	 * This test ensures that empty snapshots (no op/keyed stated whatsoever) will be reported as stateless tasks. This
+	 * happens by translating an empty {@link SubtaskState} into reporting 'null' to #acknowledgeCheckpoint.
+	 */
+	@Test
+	public void testEmptySubtaskStateLeadsToStatelessAcknowledgment() throws Exception {
```

Comment:
I think the fact that this test requires extensive whiteboxing means we should move the whole `CheckpointOperation` to a separate class and make it work independent of `StreamTask`.